### DISCUSSION TITLE
Embedding videos

### DISCUSSION THREAD

**Discussion by salmanhk** [timlrx/tailwind-nextjs-starter-blog#116] [Q&A] [answered]
Is there any way of embedding a youtube or vimeo video in the markdown files?
Thanks for the great template.

**Comment by timlrx** [comment, marked as answer]
you can just copy and paste the youtube iframe code directly to the mdx file.

![image](https://user-images.githubusercontent.com/28362229/125718511-d555a531-a171-4176-8993-d62012f1c462.png)

**Reply by 2wendyliu** [reply]
Thanks for the answer. I have a corresponding Youtube video for each of the mdx page and hope to include a new "Youtube path" field in the front matter instead of pasting the entire iframe code. In which file should I define the field and/or paste the iframe code? Thank  you!

**Reply by timlrx** [reply]
In this case, you should standardize a frontmatter field e.g. `youtube` across all your markdown documents. This will contain the youtube URL link. Since all frontmatter information is passed to the layout, you can extract it over there: https://github.com/timlrx/tailwind-nextjs-starter-blog/blob/master/layouts/PostLayout.js#L20

```
const { slug, fileName, date, title, tags, youtube } = frontMatter
```

and render the video as desired.

**Reply by 2wendyliu** [reply]
Hi Tim, 

Thanks to your kit, my website is almost done programming! However, once I downloaded the Jan. 26 snapshot, my embeded YouTube video stopped working. It was working with the Nov. 14 snapshot that I initially downloaded. The error message I got was:
This content is blocked. Contact the site owner to fix the issue.

Please advise. Once again, thanks so much for your help!

Wendy

**Reply by 2wendyliu** [reply]
const ContentSecurityPolicy = `
  default-src 'self';
  script-src 'self' 'unsafe-eval' 'unsafe-inline' giscus.app;
  style-src 'self' 'unsafe-inline' *.googleapis.com cdn.jsdelivr.net;
  img-src * blob: data:;
  media-src 'none';
  connect-src *;
  font-src 'self' fonts.gstatic.com cdn.jsdelivr.net;
  frame-src giscus.app
`
const securityHeaders = [
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
  // {
  //   key: 'Content-Security-Policy',
  //   value: ContentSecurityPolicy.replace(/\n/g, ''),
  // },
 
By the way, I commented out two lines in next.config.js as shown above and the embedded video worked again. Hope that is okay?

Wendy

**Reply by timlrx** [reply]
Yes, it's due to the new [content security policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) that we added. You can either remove it as you have done or whitelist the domain that the video is coming from e.g. youtube.com

**Reply by 2wendyliu** [reply]
Thanks for the tip. I whitelisted youtube.com instead.

**Reply by anushkaguptaaa** [reply]
Hey! How do you whitelist YouTube?
Under which sub-policy do you put it?

```js
const ContentSecurityPolicy = `
  default-src 'self';
  script-src 'self' https://www.youtube.com 'unsafe-eval' 'unsafe-inline' giscus.app;
  style-src 'self' 'unsafe-inline';
  img-src * blob: data:;
  media-src 'none';
  connect-src *;
  font-src 'self';
  frame-src giscus.app
`
```

This doesn't seem to be working for me.

**Reply by ray-android** [reply]
> Hey! How do you whitelist YouTube? Under which sub-policy do you put it?
> 
> ```js
> const ContentSecurityPolicy = `
>   default-src 'self';
>   script-src 'self' https://www.youtube.com 'unsafe-eval' 'unsafe-inline' giscus.app;
>   style-src 'self' 'unsafe-inline';
>   img-src * blob: data:;
>   media-src 'none';
>   connect-src *;
>   font-src 'self';
>   frame-src giscus.app
> `
> ```
> 
> This doesn't seem to be working for me.

I put in frame-src.

**Reply by anushkaguptaaa** [reply]
That still didn't work out for me. 
```js
const ContentSecurityPolicy = `
  default-src 'self';
  script-src 'self' 'unsafe-eval' 'unsafe-inline' giscus.app;
  style-src 'self' 'unsafe-inline';
  img-src * blob: data:;
  media-src 'none';
  connect-src *;
  font-src 'self';
  frame-src giscus.app *.youtube.com/*
`
```

Can you help me with where I am going wrong?
I even tried the following URLs
```js
youtube.com
https://www.youtube.com
*/www.youtube.com/*
*.youtube.com/embed/*
```

Am I missing any code that has be added someplace else?

**Reply by ray-android** [reply]
Are you using the no-cookie version of YouTube? If yes, then use https://www.youtube-nocookie.com/ 

If it still does not work, post your embedded YouTube code, browser console error and next.config.js ContentSecurityPolicy code.

**Reply by anushkaguptaaa** [reply]
Hey Ray, thanks for replying.

No, I'm not using no-cookie version of YouTube. Here's the embedded YouTube code
```md
<iframe
  width="560"
  height="315"
  src="https://www.youtube.com/embed/XNbtuhxNj6Q"
  title="45 Book Recommendations by @SahilShahcomedy (Part 1)"
  frameborder="0"
  allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
  allowfullscreen
></iframe>

Testing!
```

And, here's the full `next.config.js` file code (incase something else is going wrong):
```js
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})

// You might need to insert additional domains in script-src if you are using external services
const ContentSecurityPolicy = `
  default-src 'self';
  script-src 'self' 'unsafe-eval' 'unsafe-inline' giscus.app;
  style-src 'self' 'unsafe-inline';
  img-src * blob: data:;
  media-src 'none';
  connect-src *;
  font-src 'self';
  frame-src giscus.app *.youtube.com/*
`

const securityHeaders = [
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
  {
    key: 'Content-Security-Policy',
    value: ContentSecurityPolicy.replace(/\n/g, ''),
  },
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy
  {
    key: 'Referrer-Policy',
    value: 'strict-origin-when-cross-origin',
  },
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options
  {
    key: 'X-Frame-Options',
    value: 'DENY',
  },
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options
  {
    key: 'X-Content-Type-Options',
    value: 'nosniff',
  },
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-DNS-Prefetch-Control
  {
    key: 'X-DNS-Prefetch-Control',
    value: 'on',
  },
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security
  {
    key: 'Strict-Transport-Security',
    value: 'max-age=31536000; includeSubDomains',
  },
  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy
  {
    key: 'Permissions-Policy',
    value: 'camera=(), microphone=(), geolocation=()',
  },
]

module.exports = withBundleAnalyzer({
  reactStrictMode: true,
  pageExtensions: ['js', 'jsx', 'md', 'mdx'],
  eslint: {
    dirs: ['pages', 'components', 'lib', 'layouts', 'scripts'],
  },
  async headers() {
    return [
      {
        source: '/(.*)',
        headers: securityHeaders,
      },
    ]
  },
  webpack: (config, { dev, isServer }) => {
    config.module.rules.push({
      test: /\.svg$/,
      use: ['@svgr/webpack'],
    })

    if (!dev && !isServer) {
      // Replace React with Preact only in client production build
      Object.assign(config.resolve.alias, {
        'react/jsx-runtime.js': 'preact/compat/jsx-runtime',
        react: 'preact/compat',
        'react-dom/test-utils': 'preact/test-utils',
        'react-dom': 'preact/compat',
      })
    }

    return config
  },
})
```

Here's the screenshot of the browser error which says `div#sub-frame-error`

![Screenshot from 2024-06-24 20-34-44](https://github.com/timlrx/tailwind-nextjs-starter-blog/assets/76275368/1c516e19-971d-4174-bf34-32dea600da68)

**Reply by ray-android** [reply]
Replace 
`  frame-src giscus.app *.youtube.com/*
`
with
`  frame-src giscus.app https://www.youtube.com/
`

**Reply by anushkaguptaaa** [reply]
IT WORKEDDDD!!!! :partying_face: :partying_face: :partying_face: :partying_face: 
Thank you so much @ray-android 

Can you tell me why did you suggest me to put it under `frame-src` and not keep it in `script-src`?

**Reply by ray-android** [reply]
Great! I have been through this. Initially, I also put at the script src and it does not work. Then realised have to put in frame-src since we embed via iframe.

**Reply by anushkaguptaaa** [reply]
Ohhh good to know! Thanks again :)

**Reply by anushkaguptaaa** [reply]
Hey @ray-android 
Just wanted to know if you ever made the width and height of the embedded video, dynamic.
I am unable to see how I can do this.

**Reply by ray-android** [reply]
Try add a style attribute in the iframe e.g.
```
<iframe
  width="560"
  height="315"
  src="https://www.youtube.com/embed/XNbtuhxNj6Q"
  title="45 Book Recommendations by @SahilShahcomedy (Part 1)"
  frameborder="0"
  allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
  allowfullscreen
        style={{
          position: 'absolute',
          top: 0,
          left: 0,
          width: '560px',
          height: '315px,
        }}
></iframe>
```

**Reply by anushkaguptaaa** [reply]
It shows me a server error
```console
wait  - compiling /_error (client and server)...
event - compiled client and server successfully in 59 ms (416 modules)
error - Error: Build failed with 1 error:
_mdx_bundler_entry_point-abdd56ce-2b92-49c0-9ce4-0448ad8c3c39.mdx:18:29: error: [plugin: esbuild-xdm] Could not parse expression with acorn: Unexpected content after expression
error - Error: Build failed with 1 error:
_mdx_bundler_entry_point-abdd56ce-2b92-49c0-9ce4-0448ad8c3c39.mdx:18:29: error: [plugin: esbuild-xdm] Could not parse expression with acorn: Unexpected content after expression
```

**Reply by ray-android** [reply]
Try

```
<iframe
  width="560"
  height="315"
  src="https://www.youtube.com/embed/XNbtuhxNj6Q"
  title="45 Book Recommendations by @SahilShahcomedy (Part 1)"
  frameBorder="0"
  allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
  allowFullScreen
  style={{
    position: 'absolute',
    top: 0,
    left: 0,
    width: '560px',
    height: '315px',
  }}
></iframe>
```

**Reply by anushkaguptaaa** [reply]
This somewhat works but not fully xD
I'm not sure if I can even explain what this is doing. 

Overlaying on the whole website I believe.

![Screenshot from 2024-06-27 22-09-47](https://github.com/timlrx/tailwind-nextjs-starter-blog/assets/76275368/fa4641e8-2b0d-465b-9dd0-0a3116a21bd4)

**Reply by ray-android** [reply]
Use this style should fix your issue
```
  style={{
    width: '560px',
    height: '315px',
  }}
```

**Reply by anushkaguptaaa** [reply]
This does works but only for the desktop.

It looks pretty shabby when opened in mobile. How do I make it relative?

**Reply by ray-android** [reply]
The style I use is as follows. Feel free to tweak it until it suits your preferences.

```
        style={{
          position: 'absolute',
          top: 0,
          left: 0,
          width: '100%',
          height: '100%',
        }}
```

**Reply by anushkaguptaaa** [reply]
Thank you Ray!

This helped me get some direction! 
I got mine working as I want it to. I'll explain it here itself for posterity sake. 
I'd be happy to  add it as an article in this repo, if @timlrx finds it useful.

I went to `css/prism.css` and added the following code in the bottom of the file.
```css
.container {
    position: relative;
    width: 100%;
    height: 0;
    padding-bottom: 56.25%;
}
.video {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
}
```
Without making any explicit imports to the `.mdx` file, I went and enclosed the `iframe` in a `div`
Following is the tweaked code:
```html
<div class="container" align="center">
  <iframe
  width="560"
  height="315"
  src="https://www.youtube.com/embed/XNbtuhxNj6Q"
  title="45 Book Recommendations by @SahilShahcomedy (Part 1)"
  frameBorder="0"
  allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
  allowFullScreen
  class="video"
></iframe>
</div>
```

Here's the [Link](https://www.h3xed.com/web-development/how-to-make-a-responsive-100-width-youtube-iframe-embed) to the original article which gave me the idea in the first place.

Now the video embed is working perfectly in both, desktop and mobile!! Static video issue is resolved. 

# Rendered pages:

![Screenshot from 2024-06-28 00-22-58](https://github.com/timlrx/tailwind-nextjs-starter-blog/assets/76275368/6505c21b-61d1-46e2-b4bc-1bd085e3ceeb)

![Screenshot from 2024-06-28 00-23-25](https://github.com/timlrx/tailwind-nextjs-starter-blog/assets/76275368/3f6a64ea-4400-4dac-9ceb-4c386000b26b)

Hoping this would help someone in the future! :partying_face: :sparkles:

**Comment by 2wendyliu** [comment]
Hello Timothy!

Nice to hear from you. And thank you so much for making the code available.
I am planning to use it to build a band site with data/music, data/blog and
data/memberbio to replace my old and buggy Drupal 7 site. This feels like
an impossible task as I am new to github, nextjs, tailwind, etc. But I love
the idea of having my blog and songs stored in mdx with nestable paths.
Your code on GitHub gave me a place to start. My next challenge is to
modify the code to fit my directory structure. Do you mind pointing out the
files to modify? Thank you so so much!

Wendy

**Reply by timlrx** [reply]
Hi Wendy, yes it will be quite challenging if you are new to react / nextjs. You probably won't want to store the music and songs with the website, but use a file sharing service like youtube. Each mdx file in `data/music` can then contain a frontmatter link to the song as described above. As to the files to modify, for each folder / section you include in data, you probably want a page for it. You can copy the `pages/blog` example and modify it so you get a `page/music`. 

You will then need to modify them to source the content from the right folder, by modifying `blog` to the appropriate folder name e.g. https://github.com/timlrx/tailwind-nextjs-starter-blog/blob/master/pages/blog/%5B...slug%5D.js#L10

You would probably want to customize and use a different template for each page type. You can add a new layout to the `layouts` folder and modify the `DEFAULT_LAYOUT` field over [here](https://github.com/timlrx/tailwind-nextjs-starter-blog/blob/master/pages/blog/%5B...slug%5D.js#L7)

**Comment by 2wendyliu** [comment]
Great. I will follow your instructions to give it a try. How should I
modify components/Pagination.js as it contains "/blog/"

Thanks,

Wendy
